### ISSUE TITLE
--models unrecognised option

### ISSUE THREAD

**Issue by PDSc0tt** [jts/nanopolish#88]
Hi Jared,
I've been going round in circles over the last few days trying to nanopolish sequencing data and getting nowhere (not helped by my lack of Linux skills). I was initially using a build from July and Nick Loman suggested I remove it and do a clean install of the current version of nanopolish.
However, despite installing the recent version, I am having the same issue, regardless of whether I use the workflow in the Nanopolish GitHub instructions, or the poretools approach I learned from Josh in July.
When I get the sequence data to the point of doing eventalign, I get the following issue:

From my poretools approach:
ubuntu@hivp12:~/p12$ nanopolish eventalign -t 8 --sam -r p12NB01.fa -b p12NB01.sorted.bam -g reference.fa --models nanopolish_models.fofn | samtools view -Sb - | samtools sort -f - p12NB01.eventalign.sorted.bam samtools index p12NB01.eventalign.sorted.bam
[bam_header_read] EOF marker is absent. The input is probably truncated.
eventalign: unrecognized option '--models'
[samopen] no @SQ lines in the header.
[sam_read1] missing header? Abort!

Also, if I follow the nanopolish instructions on GitHub (using reads as my sequence filenames), I get the same result:
ubuntu@hivp12:~/p12$ nanopolish eventalign -t 8 --sam -r reads.fa -b reads.sorted.bam -g reference.fa --models nanopolish_models.fofn | samtools view -Sb - | samtools sort -f - reads.eventalign.sorted.bam samtools index reads.eventalign.sorted.bam
[bam_header_read] EOF marker is absent. The input is probably truncated.
eventalign: unrecognized option '--models'
[samopen] no @SQ lines in the header.
[sam_read1] missing header? Abort!

I have no idea why the --models command is failing. Josh was not sure what could be going wrong either. It may be that I'm just not doing something right, due to my lack of experience doing MinION nanopolish work. Any assistance would be greatly appreciated.
Thanks
Paul

**Comment by jts** [comment]
Hi Paul,

I'm happy to help - I'm pretty confident though that you are still running an old version of nanopolish as the `--models` flag is present in the current version. Could you go into the directory you cloned from github and run `git log --pretty=oneline | head -1` and paste the result? If your code is indeed up-to-date you might want to check to make sure that the correct path to nanopolish is being picked up by looking at the output of `which nanopolish`.

Cheers,
Jared

**Comment by PDSc0tt** [comment]
Hi Jared,
Thanks for the response. The results of git log --pretty=oneline | head -1 is:
6406f78c01d05f968bc841c03e4ccb4ea208677a when -w is not specified, polish/call on the whole genome, closes #84

Which nanopolish shows the following (from the nanopolish directory):
/usr/bin/nanopolish

Hopefully I have done the correct thing and the results are interpretable.
Thanks again for your help.
Paul

**Comment by jts** [comment]
I think the problem is that the previous version of nanopolish you installed (in `/usr/bin`) is old and getting picked up. I suggest deleting that version and giving the full path to the version you cloned from github:

```
/path/to/nanopolish/nanopolish eventalign [...]
```

**Comment by PDSc0tt** [comment]
OK now it is slightly further along. I'm not getting the --models issue any more, but I'm getting the following after performing all the steps indicated in the workflow on the nanopolish site, so I can't see if it's working correctly:

../software/nanopolish/nanopolish eventaligh -t 8 --sam -r reads.fa -b reads.sorted.bam -g reference.fa --models nanopolish_models.fofn | samtools view -Sb - | samtools sort -f - reads.eventalign.sorted.bam samtools index reads.eventalign.sorted.bam
[bam_header_read] EOF marker is absent. The input is probably truncated.
[samopen] no @SQ lines in the header.
[sam_read1] missing header? Abort!

Do I need to use poretools for the MinION 2d data to work through fastq files as well as fasta files? I feel it's getting close now, but I'm obviously not doing something right.
Thanks again
Paul

**Comment by jts** [comment]
I suggest try running eventalign on its own without piping into samtools and seeing if the output looks OK (is it valid `sam`?). I suspect there is something wrong with your input bam file `read.sorted.bam`

**Comment by PDSc0tt** [comment]
I think you're right. I ran eventalign on its own and got the following result:

../software/nanopolish/nanopolish eventalign -t 8 --sam -r reads.fa -b reads.sorted.bam -g reference.fa --models nanopolish_models.fofn
registering model t.007-base
registering model t.007-ONT
nanopolish: src/alignment/nanopolish_eventalign.cpp:867: int eventalign_main(int, char**): Assertion `bam_idx != NULL' failed.
Aborted (core dumped)

I'll try doing the old pore tools route to see if I can force a valid bam file.
Thanks
Paul

**Comment by jts** [comment]
No, just run `samtools index reads.sorted.bam` and you should be OK

**Comment by PDSc0tt** [comment]
That seemed to do the trick. When I'm doing the next step in parallel, do I need the full path to nanopolish for the make range.py (i.e. python ../software/nanopolish nanopolish_makerange.py...)?
Thanks and sorry for the naive questions.
Paul

**Comment by jts** [comment]
Yes, you'll need the full path - no problem, ask any questions you have.

**Comment by PDSc0tt** [comment]
Thanks. I had to try different iterations to get the paths right in both instances, but in the end I obtained the following error when inputting the command as shown:

python ../software/nanopolish/scripts/nanopolish_makerange.py reference.fa | parallel --results nanopolish.results -P 8 \ ../software/nanopolish/nanopolish variants --consensus polished.{1}.fa -w {1} -r reads.fa -b reads.sorted.bam -g reference.fa -e reads.eventalign.sorted.bam -t 4 --min-candidate-frequency 0.1 --models nanopolish_models.fofn
When using programs that use GNU Parallel to process data for publication please cite:

  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,
  ;login: The USENIX Magazine, February 2011:42-47.

This helps funding further development; and it won't cost you a cent.
Or you can get GNU Parallel without this requirement by paying 10000 EUR.

To silence this citation notice run 'parallel --bibtex' once or use '--no-notice'.

#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
registering model t.007-base
registering model t.007-ONT
nanopolish: src/alignment/nanopolish_alignment_db.cpp:337: BamHandles _initialize_bam_itr(const string&, const string&, int, int): Assertion `bam_idx != NULL' failed.

Have I missed something? Do I need to replace {1} with an alternative value? I may have to break the pipeline into individual sections, but I'm not really sure where to start with this, but I know that I'm getting closer.
Paul

**Comment by jts** [comment]
run: `samtools index reads.eventalign.sorted.bam`

**Comment by PDSc0tt** [comment]
Seems to be working. Thank you so much for helping me out. I'll let you know how I got on.
Cheers
Paul

**Comment by jts** [comment]
Great!

**Comment by PDSc0tt** [comment]
Hi Jared,
That worked a treat. Thank you. I do have one other question though on a related note.
Once eventalign and variants pipelines are completed, is there a way to generate a consensus sequence from the MinIOn reads? The reason I ask is that I have reads for a 1.3kb amplicon and want to generate a consensus sequence from the MinION data, as well as knowing the variants from a reference sequence that I was able to do last night with your help. 
I remember using consensus_make before in the older version, but I cannot recall the pipeline needed to use this (I remember using poretools and doing consensus_make and consensus preprocessing to generate reads.pp.fa files) and I wanted to check that it was still an option in the current version of nanopolish before I go barging in. As demonstrated yesterday, my command line computing skills are not very good and I don't want to make a mess of the data analysis.
Thanks for your continued help.
Paul

**Comment by PDSc0tt** [comment]
Hi Jared,
It appears that a new issue has developed after successful eventalign analysis on 6 samples. In all instances I'm now getting an error when eventalign is processing regarding the nanopolish alphabet for amiguities and it occurs on different CLIMB servers with separate installations. Any ideas what the issue may be, as I cannot figure it out and it works when I repeat the analysis on any of the 6 I have already done?
Thanks
Paul

../software/nanopolish/nanopolish eventalign -t 8 --sam -r p12NB01.fa -b p12NB01.sorted.bam -g p12Aref.fa --models nanopolish_models.fofn | samtools view -Sb - | samtools sort -f - p12NB01.eventalign.sorted.bam samtools index p12NB01.eventalign.sorted.bam
[bam_header_read] EOF marker is absent. The input is probably truncated.
registering model t.007-base
registering model t.007-ONT
[fai_load] build FASTA index.
nanopolish: ./src/common/nanopolish_alphabet.h:173: virtual std::__cxx11::string Alphabet::disambiguate(const string&) const: Assertion `IUPAC::isValid(out[i])' failed.
[samopen] SAM header is present: 1 sequences.
[sam_read1] reference 'ID:bwa   PN:bwa  VN:0.7.12-r1039 CL:bwa mem -x ont2d -t 8 p12Aref.fa p12NB01.fa
' is recognized as '*'.
[main_samview] truncated file.

**Comment by jts** [comment]
Hi,

Did you remove the ambiguous characters from your input genome sequences?

jared

**Comment by PDSc0tt** [comment]
Hi Jared,
This solved the issue. Is there a way to leave ambiguities in? or can Nanopolish not use these at the current time? Is there a way to calculate the basecall accuracy and where I can find the quality filter level in the scripts for variant calling (I'm still learning how to do a lot of this type of analysis).
Thanks again for your help.

**Comment by jts** [comment]
Eventually I'll modify the code to deal with ambiguities but it is currently not possible. 

I use https://github.com/arq5x/nanopore-scripts to calculate basecall accuracy. When filtering variants I look at the `SupportFraction` field (closer to 1 is better) and the `QUAL` score (higher is better). Guidance on how to filter variants is still in development and I'd be interested to hear what works for you.
